batchflow/utils_transforms.py

```diff
@@ -0,0 +1,217 @@
+        """ Denormalize image with provided stats.
+
+        Parameters
+        ----------
+        array : numpy.ndarray
+
+        normalization_stats : dict or None, optional
+            If provided, then used to get statistics for denormalization, by default None.
+            If None, self.normalization_stats will be used.
+        return_stats : bool, optional
+            Whether to return stats used for normalization, by default False
+
+        Returns
+        -------
+        numpy.ndarray or (numpy.ndarray, dict)
+        """
+        array = array if self.inplace else array.copy()
+        mode = self.mode if mode is None else mode
+
+        if self.normalization_stats is not None:
+            normalization_stats = self.normalization_stats
+
+        if callable(mode):
+            array[:] = mode(array, normalization_stats)
+        else:
+            if 'std' in mode:
+                array *= normalization_stats['std']
+            if 'mean' in mode:
+                array += normalization_stats['mean']
+            if 'min' in mode and 'max' in mode:
+                if self.clip_to_quantiles:
+                    min_, max_ = normalization_stats['q']
+                else:
+                    min_, max_ = normalization_stats['min'], normalization_stats['max']
+
+                if min_ != max_:
+                    array *= max_ - min_
+                array += min_
+
+        return array
+
+class Quantizer:
+    """ Class to hold parameters and methods for (de)quantization.
+
+    Parameters
+    ----------
+    ranges : tuple
+        Bounds to create bins.
+    clip : bool, optional
+        Whether to clip data to selected ranges, by default True
+    center : bool, optional
+        Whether to make data have 0-mean before quantization, by default False
+    mean : _type_, optional
+        Mean value for centering, by default None
+    dtype : numpy.dtype, optional
+        dtype for the quantized array, by default np.int8
+    copy : bool, optional
+        Whether to make copy of the data under the hood, by default False.
+        Enabled copy will not allow to change input data but quantization will be slower.
+    """
+    def __init__(self, ranges, clip=True, center=False, mean=None, dtype=np.int8, copy=False):
+        if dtype not in [np.int8, np.int16, np.uint8, np.uint16]:
+            raise TypeError(f'{dtype} is not supported, use int8, int16, uint8, uint16')
+        self.ranges = ranges
+        self.clip, self.center = clip, center
+        self.mean = mean
+        self.dtype = dtype
+        self.copy = copy
+
+        n_bins = np.iinfo(dtype).max - np.iinfo(dtype).min - 1
+        self.bins = np.histogram_bin_edges(None, bins=n_bins, range=ranges).astype(np.float32)
+
+    def quantize(self, array):
+        """ Quantize data: find the index of each element in the pre-computed bins and use it as the new value.
+        Converts `array` to `self.dtype`. Lossy.
+        """
+        if self.copy:
```

Comment:
looks like `copy` should be a parameter of this function, not of the instance itself